### ISSUE TITLE
Class weighting in tf.losses.softmax_cross_entropy

### ISSUE THREAD

**Issue by w4nderlust** [tensorflow/tensorflow#10021]
Feature request

In `tf.losses.softmax_cross_entropy` there's an optional field `weights`. I assumed this field was used for assigning a different weight to each class, but it actually is used to assign a weight to each sample in the batch. In my use case i have a `batch_size` of `128` and `30` classes, so I was passing a `[1, 30]` tensor and got this error:

```
InvalidArgumentError (see above for traceback): Incompatible shapes: [128] vs. [30]
	 [[Node: optimizer/gradients/cross_entropy/softmax_cross_entropy_loss/Mul_grad/BroadcastGradientArgs = BroadcastGradientArgs[T=DT_INT32, _device="/job:localhost/replica:0/task:0/gpu:0"](optimizer/gradients/cross_entropy/softmax_cross_entropy_loss/Mul_grad/Shape, optimizer/gradients/cross_entropy/softmax_cross_entropy_loss/Mul_grad/Shape_1)]]
```

I looked at the implementation, confirmed that the function expects `batch_size` as the dimension of thensor and realized that my expected behavior cannot be achieved easily as `tf.nn.softmax_cross_entropy_with_logits` doesn't have a weight parameter.

My current workaround solution is to re-implement this function calculating the loss for each class and then multiplying for the weight vector that I pass, but that's inefficient compared to the optimized implementation of `tf.nn.softmax_cross_entropy_with_logits`.

So my request is:
- provide an optimized `tf.nn.softmax_cross_entropy_with_logits` that also accepts `weights` for each class as a parameter
- use it inside `tf.losses.softmax_cross_entropy` so that one can pass weights as a scalar, a `[batch_size, 1]` tensor, a `[1, num_classes]` tensor or a `[batch_size, num_classes]` tensor (the same dimension of  `onehot_labels`)

**Comment by ali01** [comment]
Thanks for the suggestion @w4nderlust. Marking this as "Contributions Welcome" for now.

**Comment by foxik** [comment]
Personally I am opposed to such an extension -- the ``weights`` argument of ``tf.losses.softmax_cross_entropy`` can easily achieve class weighting (by using `tf.gather` on the class weights and class indices), using one additional line of code.

Yes, the `tf.gather` requires additional time and space, but linear in size of the output layer -- most networks spend several orders of magnitude more time in the computations before the output layer, so the effect of the suggested optimalization is minor (while requiring API change).

**Comment by w4nderlust** [comment]
@foxik my current solution is:
```
sample_weights = tf.reduce_sum(tf.multiply(self.one_hot_labels, class_weights), 1)
```
which is similar to `tf.gather`, but both are workarounds. I believe that allowing also for class weighting in the loss is a more general solution.

**Comment by foxik** [comment]
@w4nderlust I see it the opposite way -- the current solution is a more general one. It allows class weighting as a simple case, but also allows weighting the instances individually (this is sometimes needed if you either are given weights beforehand, or if you for example want the same sum of positive/negative examples to be the same, for example).

**Comment by w4nderlust** [comment]
@foxik when I say general I refer to the interface, not the capabilities.

My proposed solution allows both class and instance weighting, so it allows exactly the same capabilities that are available now, but in simpler and more straightforward way, without requiring workarounds.

Moreover, what's different in my proposal is the interface, it would become more general if it would accept all possible formats (scalar, [batch_size], [class_size] and [batch_size, class_size]) instead of accepting only [batch_size] (and I guess also scalar, haven't tried it) like it does now.

**Comment by foxik** [comment]
I am not sure what the meaning of `[batch_size, class_size]` is (because there are only `batch_size` instances, but you are giving `batch_size * class_size` weights).

From the API point of view, my personal oppinion is that you should not mix both instance-based weights and class-based weights in one argument -- I would instead keep `weights` as it is and add `class_weights` argument, if I were to proceed with the request.

However, when designing APIs, there is a cost for extending it -- too rich API methods and arguments are not always better (simplicity is also an important factor). As a rule of thumb, if required functionality can be achieved by a line of code without harming performance, I personally do not believe the API should be extended (unless the functionality is used by everyone all the time).

But I respect that you have another oppinion -- lets wait for someone else to step in.

**Comment by w4nderlust** [comment]
@foxik [batch_size, class_size] is the size of both the logits and the onehot_labels we already passing to the softmax. I believe you should be able to pass weights with the same size. and if they happen to be just [bastch_size] then they should be broadcasted to each class, or if they are only [class_size] they should be broadcasted to each example in the batch, and finally if you pass a scalar it should be broadcasted in both batch_size and class_size. Hope this is more clear now.

Adding an additional class weights parameter would be ok, but if you multiply the class weights and the batch weight parameter you get the [bach_size, class_size] weight matrix I'm talking about. Either way would be fine IMHO.

I believe that an optional parameter, if useful, does no big harm and doesn't make the API less simple, it may actually make the model's code slightly simpler. Anyway, I understand that the effort to implement this may not be worth both the rather small advantage that it would give, but I'm proposing a feature I'd like to have, it's not up to me to decide how important or urgent it is :) I respect your point of view too anyway.

**Comment by shaibagon** [comment]
@w4nderlust  implementing [`"info gain"](http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1InfogainLossLayer.html#details) loss should give you the ability to have class weights as a special case of the more general infogain loss.

**Comment by JackieLeeTHU11** [comment]
@w4nderlust This solution works well!

**Comment by w4nderlust** [comment]
@shaibagon: I already have something that works, but having something like that implemented inside tf would me more elegant I guess.

@JackieLeeTHU11: thanks for the suggestion. I think you are computing something different anyway. Multiplying the weights for the one hots you're asking your model to predict those weights and penalize it if doesn't. In my case I still want the model to predict ones and zeros, but I want to weight the loss for some classes lower.

**Comment by dickreuter** [comment]
In my loss function I would like to weight each sample differently and in each sample, each category should be weighted differently as well depending on the outcome. Meaning if in a cross entropy the one_hot is correctly specified, a different weight needs to be applied than when the output is incorrect. So I would need two weights per category. A tensor with rank 3. One dimension for the samples, a second dimension for the amount of classes (let's assume just one output neuron for simplicity), and a third dimension that differentiates between correct and incorrect match for that one binary output neuron.

Is there any way to do that with one of the above loss functions?

**Comment by zihuaweng** [comment]
https://stackoverflow.com/a/46984951/8906149
This answer works fine for me~~

**Comment by klyan** [comment]
@zihuaweng ，Hi zihua, 同样样本加权的问题，不知道如何将sample_weight传给custom model,感谢！ https://stackoverflow.com/questions/49312839/how-to-set-parts-of-positive-samples-weight-in-tensorflow-for-binary-classficati

**Comment by wt-huang** [comment]
This can be done in keras.

**Comment by w4nderlust** [comment]
If anything, that looks more like an additional reason to have it in the lower level APIs

**Comment by bipinkc19** [comment]
Currently while implementing weighted loss there is tf.nn.weighted_cross_entropy_with_logits which uses sigmoid so it is restricted to binary classification. For multiclass classification using softmax categorical cross entropy, there isn't any feature.

Also in many forums, while implementing the class weights people often suggest something like this:

weighted_logits = tf.multiply(weights, logits)
loss = tf.nn.softmax_cross_entropy_with_logits_v2

Is this the correct method? Because we are messing around with the output of the model, while we should have been messing around with the calculated [batch_size, num_class] shape loss tensor.

I think this is the correct way:
        y_hat_softmax = tf.nn.softmax(logits)
        y_cross = y * tf.log(y_hat_softmax)
        weighted_cross = tf.multiply(class_weight, y_cross)
        weighted_cross_reduced = - tf.reduce_sum(weighted_cross, 1)
        loss = tf.reduce_mean(weighted_cross_reduced)

Here we are changing the y_cross which should be actually changed, which then changes the gradient flow. But changing logits (by multiplying with weight) will make the model understand that the correct output is different thus biasing result.

eg. If a model has three classes and the output logit is [1, 4, 2] then it is making the correct decision if the actual class is the second one. But multiplying with weight [1, 1, 4] will make the result [1, 4, 8] thus saying that the third logit should have been high to the loss function, thus completely changing the supervision in supervised learning. But actually, we should have been messing around with the loss by scaling the loss according to the weights. 

Could you guys verify this?

**Comment by xlambein** [comment]
@bipinkc19 If you do the maths, you'll see that class weighting is strictly equivalent to sample weighting.  Cross entropy loss for a given sample with class `c` is defined as (in pseudocode):

```loss = sum(w[i] * y[i] * log(p[i]))```

where `w[i]` is the weight for the `i`th class, `y` is the one-hot encoding of `c`, and `p` is the softmax probabilities.  If you consider that the sample of that class is `c`, then you know that `y[c] = 1`, and `y[i] = 0` for `i != c`.  Hence:

```loss = w[c] * log(p[i])```

All the other class weights have vanished.  The loss (and thus, its gradient) only depend on `w[c]`, and so class weights can be expressed as sample weights.

**Comment by w4nderlust** [comment]
@xlambein you are assuming that all datapoints from the same class should be weighted the same, which is not true in the most general case. Your math should be: `loss = sum(w_s[i] * w_c[y[i]] * y[i] * log(p[i]))` where `w_s[i]` is the weight of the i-th datapoint and `w_c[y[i]]` is the weight of the class of the i-th datapoint.
As already discussed several times in this thread you can combine class and sample weight together before providing them to the current api, but that doesn't need to be the case, the multiplication could be done inside the provided api loss computation. It's a matter of api design.

**Comment by xlambein** [comment]
@w4nderlust I think you did not understand the purpose or the content of my post.  I don't have an opinion about API design.  I was just pointing out that a "weighted cross entropy", as suggested by @bipinkc19, actually reduces to sample weights, since all but one weight end up disappearing from the expression.  Hence, the implementation they suggested is not more expressive than the current implementation in TensorFlow (and is slower, since it's not optimized).
If you read carefully, you'll see that I'm specifically talking about "loss for a given sample".  When I iterate over `i` in my sum, I iterate over the classes, not the data points.  If you generalize over multiple samples, you need to wrap that sum into another sum, but that doesn't change the validity of my statement.
As for what you are talking about: as you said, it's a matter of API design.

**Comment by bipinkc19** [comment]
@xlambein @w4nderlust Thanks for clearing out that class weights by the method mentioned by me converts to sample weights. But it still doesn't explain why most implementation are done this way:
```weighted_logits = tf.multiply(weights, logits)
loss = tf.nn.softmax_cross_entropy_with_logits_v2```
Because, we are messing around with the output logits and changing its scalar value. 
1. Isn't the method of class weight to sample weight okay to do?
2. Even when we want the class weights to apply to all other classes like you said by multiplying it with the output logits. We are changing the ouput value for the model. I think it might disrupt the whole process like.
If a model has three classes and the output logit is [1, 4, 2] then it is making the correct decision if the actual class is the second one. But multiplying with weight [1, 1, 4] will make the result [1, 4, 8] thus saying that the third logit should have been high to the loss function, thus completely changing the supervision in supervised learning. But actually, we should have been messing around with the loss by scaling the loss according to the weights.

Could you guys also mention how you do class weights implementation. :)

**Comment by w4nderlust** [comment]
@bipinkc19 I agree with you, weighting should be applied after the softmax to influence the gradients rather than at the logits / output level. In my implementation I multiply after computing the cross entropy and before aggregating (so the output of the cross entropy is not reduced, I multiply by weights for each datapoint based on class and specific datapoint, and finally I sum and divide by the batch size.

@xlambein I understood both the purpose and the content of your post. The fact that the provided weighting mechanism is as expressive as needed for both class and sample weighting was clear from the beginning of this discussion. The discussion is about what is the best api design. As you said, multiplying outside the `softmax_cross_entropy_with_logits` function is non optimized, so a better interface providing a parameter for both weighting schemes could encapsulate a more optimized implementation imho.

**Comment by xlambein** [comment]
@w4nderlust Maybe it was clear for you, but it wasn't for me.  I encountered a situation where I wanted to weight my loss by a `[batch_size, n_classes]` tensor, and found this thread.  My first intuition was to think the current implementation was limited in expressiveness, and that intuition was compounded by the fact that you suggested a richer API.  Only when I did the maths myself did I realize it wasn't the case.  Hence, since @bipinkc19 seemed to have the same problem as me, I thought it would be useful to share the short math that explains why `[batch_size, 1]` weights are just as expressive as `[batch_size, n_classes]` weights.

@bipinkc19 I would do the loss with class weight the way it's suggested at the start of the thread, with `tf.gather` (not tested):
```python
class_weights = tf.constant([1., 2., 3., 4.])  # 4 classes
sample_weights = tf.gather(class_weights, labels)
loss = tf.losses.sparse_softmax_cross_entropy(labels, logits, weights=sample_weights)
```
Or alternatively, with one-hot labels, you can use matrix multiplication to compute the sample weights (multiply `[batch_size, n_classes]` one-hot labels with `[n_classes, 1]` class weights to get `[batch_size, 1]` sample weights), and compute the loss with `tf.losses.softmax_cross_entropy`.

I'm not sure I understand what you're asking in (1).  As for (2), I think you're right that it's much better to change the loss than to mess with the logits.  That being said, I'm not sure how bad it would actually be.  In the case of your example, your model might just learn to output a smaller logit for the third class.  I'm too much of a novice to understand the impact of that.
